### ISSUE TITLE
[Bug]: vllm gives 400 bad request with high logprob count on Mistral-small because pydantic flags logits of byte tokens as invalid

### ISSUE THREAD

**Issue by JD-P** [vllm-project/vllm#16540]
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
PyTorch version: 2.6.0+cu124                                                                                                           
Is debug build: False                                                                                                                  
CUDA used to build PyTorch: 12.4                                                                                                       
ROCM used to build PyTorch: N/A                            
                                                                   
OS: Debian GNU/Linux 11 (bullseye) (x86_64)       
GCC version: (Debian 10.2.1-6) 10.2.1 20210110                                                                                         
Clang version: Could not collect                                                                                                                                                                                                                                               CMake version: version 3.18.4                                                                                                          
Libc version: glibc-2.31                          
                                                                   
Python version: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) [GCC 12.3.0] (64-bit runtime)                                                                                                                                                                
Python platform: Linux-5.10.0-33-cloud-amd64-x86_64-with-glibc2.31                                                                                                                                                                                                             
Is CUDA available: True                                                                                                                                                                                                                                                        
CUDA runtime version: 12.2.128     
CUDA_MODULE_LOADING set to: LAZY                                                                                                       
GPU models and configuration:                                                                                                          
GPU 0: NVIDIA H100 80GB HBM3                                                                                                           
GPU 1: NVIDIA H100 80GB HBM3                                                                                                           
GPU 2: NVIDIA H100 80GB HBM3                                      
GPU 3: NVIDIA H100 80GB HBM3                                                                                                           
GPU 4: NVIDIA H100 80GB HBM3                                                                                                           
GPU 5: NVIDIA H100 80GB HBM3                                                                                                           
GPU 6: NVIDIA H100 80GB HBM3                                                                                                           
GPU 7: NVIDIA H100 80GB HBM3                                                                                                           
                                                                                                                                       
Nvidia driver version: 550.54.15                                  
cuDNN version: Could not collect                                                                                                       
HIP runtime version: N/A                                                                                                               
MIOpen runtime version: N/A                                                                                                            
Is XNNPACK available: True                                                                                                             
                                                                                                                                       
CPU:                                                                                                                                   
Architecture:                         x86_64                      
CPU op-mode(s):                       32-bit, 64-bit                                                                                   
Byte Order:                           Little Endian                                                                                    
Address sizes:                        52 bits physical, 57 bits virtual    
CPU(s):                               208                                                                                              
On-line CPU(s) list:                  0-207                                                                                            
Thread(s) per core:                   2                                                                                                
Core(s) per socket:                   52                                                                                               
Socket(s):                            2                                                                                                                                                                                                                                        
NUMA node(s):                         2                                                                                                                                                                                                                                        
Vendor ID:                            GenuineIntel                                                                                                                                                                                                                             
CPU family:                           6                                                                                                
Model:                                143                                                                                              
Model name:                           Intel(R) Xeon(R) Platinum 8481C CPU @ 2.70GHz                   
Stepping:                             8                                                                                                
CPU MHz:                              2699.998                                                                                         
BogoMIPS:                             5399.99                                                                                          
Hypervisor vendor:                    KVM                                                                                              
Virtualization type:                  full                                                                                                                                                                                                                                     
L1d cache:                            4.9 MiB                                                                                          
L1i cache:                            3.3 MiB                                                                                          
L2 cache:                             208 MiB                                                                                          
L3 cache:                             210 MiB                                                                                          
NUMA node0 CPU(s):                    0-51,104-155                                                                                     
NUMA node1 CPU(s):                    52-103,156-207                                                                                   
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected                                                                                                                                                                                                                             
Vulnerability L1tf:                   Not affected                                                                                     
Vulnerability Mds:                    Not affected                                                                                     
Vulnerability Meltdown:               Not affected                                                                                     
Vulnerability Mmio stale data:        Not affected                                                                                     
Vulnerability Reg file data sampling: Not affected         
Vulnerability Retbleed:               Not affected                 
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp                              
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization                                                                                                                                                                     Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence    
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 p
cid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rtm avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx
512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 arat avx512vbmi umip avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid cldemote movdiri movdir64b fsrm md_clear serialize tsxldtrk arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-ml-py==12.560.30
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] pyzmq==26.2.0
[pip3] torch==2.6.0
[pip3] torch-dist-utils==0.1.0
[pip3] torchaudio==2.6.0
[pip3] torchvision==0.21.0
[pip3] transformers==4.51.2
[pip3] triton==3.2.0
[conda] numpy                     1.25.2                   pypi_0    pypi
[conda] nvidia-ml-py              11.495.46                pypi_0    pypi
[conda] pyzmq                     24.0.1          py310h330234f_1    conda-forge
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.8.3
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
        GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      NV18    NV18    NV18    NV18    NV18    NV18    NV18    0-51,104-155    0               N/A
GPU1    NV18     X      NV18    NV18    NV18    NV18    NV18    NV18    0-51,104-155    0               N/A
GPU2    NV18    NV18     X      NV18    NV18    NV18    NV18    NV18    0-51,104-155    0               N/A
GPU3    NV18    NV18    NV18     X      NV18    NV18    NV18    NV18    0-51,104-155    0               N/A
GPU4    NV18    NV18    NV18    NV18     X      NV18    NV18    NV18    52-103,156-207  1               N/A
GPU5    NV18    NV18    NV18    NV18    NV18     X      NV18    NV18    52-103,156-207  1               N/A
GPU6    NV18    NV18    NV18    NV18    NV18    NV18     X      NV18    52-103,156-207  1               N/A
GPU7    NV18    NV18    NV18    NV18    NV18    NV18    NV18     X      52-103,156-207  1               N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

VLLM_ALLOW_LONG_MAX_MODEL_LEN=1
LD_LIBRARY_PATH=/home/ext_johndavidpressman_gmail_com/minihf/env_minihf/lib/python3.10/site-packages/nvidia/nvjitlink/lib:/usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/cuda/lib64:/usr/local/nccl2/lib:/usr/local/cuda/extras/CUPTI
/lib64
NCCL_CUMEM_ENABLE=0
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
CUDA_MODULE_LOADING=LAZY
```

</details>


### 🐛 Describe the bug

Unfortunately just starting up the vllm server and sending a request that causes the bug doesn't seem to reproduce it, yet if I run my ReAct LLM agent for a while I will eventually reach a state where pydantic flags one of the logits returned by the mock OpenAI API as invalid for the schema. I assume what's going on here is someone set pydantic to flag anything that's not a string as an invalid token and it turns out that some models like Mistral-small can return partial UTF-8 as bytes, breaking it.

I am running VLLM with these settings:

```
python -m vllm.entrypoints.openai.api_server --model "mistralai/Mistral-Small-3.1-24B-Instruct-2503" --served-model-name "mistralai/Mistral-Small-3.1-24B-Instruct-2503" --max-logprobs 100 --gpu-memory-utilization=0.9 --disable-log-requests --disable-log-stats --port 5001 --tensor-parallel-size 8 --max-num-seqs 512 --enable-prefix-caching --max-model-len 131072 --tokenizer_mode mistral --config_format mistral --load_format mistral --tool-call-parser mistral
```

And my error looks like this:

```
With logprobs = 100
{'object': 'error', 'message': "1 validation error for CompletionLogProbs\ntop_logprobs.0.b' \\xe2'.[key]\n  Input should be a valid string, unable to parse raw data as a unicode string [type=string_unicode, input_value=b' \\xe2', input_type=bytes]\n    For further information visit https://errors.pydantic.dev/2.10/v/string_unicode", 'type': 'BadRequestError', 'param': None, 'code': 400}

With logprobs = 0
{'id': 'cmpl-651627d6c7d043f9b4a5deb6ac900932', 'object': 'text_completion', 'created': 1744484366, 'model': 'mistralai/Mistral-Small-3.1-24B-Instruct-2503', 'choices': [{'index': 0, 'text': ' Yes', 'logprobs': {'text_offset': [0], 'token_logprobs': [-0.759935736656189], 'tokens': [' Yes'], 'top_logprobs': [{' Yes': -0.759935736656189}]}, 'finish_reason': 'length', 'stop_reason': None, 'prompt_logprobs': None}], 'usage': {'prompt_tokens': 72071, 'total_tokens': 72072, 'completion_tokens': 1, 'prompt_tokens_details': None}}

With no unicode in input prompt string
{'object': 'error', 'message': "1 validation error for CompletionLogProbs\ntop_logprobs.0.b' \\xe2'.[key]\n  Input should be a valid string, unable to parse raw data as a unicode string [type=string_unicode, input_value=b' \\xe2', input_type=bytes]\n    For further information visit https://errors.pydantic.dev/2.10/v/string_unicode", 'type': 'BadRequestError', 'param': None, 'code': 400}
```

When I reach the bugged state of the server this code is sufficient to reproduce the issue (external file for the .json because it's 70k characters and I'm not sure how sensitive the formatting is to reproduce the issue so I just uploaded the file as is to the site for my agent framework, I've also added it as an attachment to this issue):

```
import random
import json
import requests

prompts = requests.get("https://minihf.com/vllm_utf8_logprobs_error_reproduce_prompts.json").json()

port = 5001
n = 1
model_name = "mistralai/Mistral-Small-3.1-24B-Instruct-2503"

payload = {
    "n": n,
    "temperature": 1,
    "top_k": 50,
    "repetition_penalty": 1.02,
    "max_tokens": 1,
    "model": model_name,
    "prompt": prompts,
    "stream": False,
    "logprobs": 100,
    "seed": random.randrange(1000000)
}

print("With logprobs = 100")
print(requests.post(f"http://localhost:{port}/v1/completions/", json=payload).json(), end="\n\n")

payload["logprobs"] = 0
print("With logprobs = 0")
print(requests.post(f"http://localhost:{port}/v1/completions/", json=payload).json(), end="\n\n")

no_unicode = prompts[0].replace("\u2019", "'").replace("\U0001f642", ":)").replace("\u201c", '').replace("\u201d", '"').replace("\u2014", "-")
assert no_unicode.encode("ascii")
prompts2 = [no_unicode,]
payload["logprobs"] = 100
payload["prompts"] = prompts2
print("With no unicode in input prompt string")
print(requests.post(f"http://localhost:{port}/v1/completions/", json=payload).json(), end="\n\n")
```

I don't think it's actually any feature of the text itself that causes the problem because taking the number of logodds returned by the API to zero and sending the same prompt doesn't cause an issue. This implies that it is what I previously described with pydantic having had its server side validation incorrectly specified for what is allowed to be a token.

This is what it looks like on the server side when it throws the error, setting my log level to debug does not seem to have helped:

```
INFO:     127.0.0.1:58276 - "POST /v1/completions/ HTTP/1.1" 307 Temporary Redirect
DEBUG 04-12 18:59:26 [core.py:420] EngineCore loop active - local unfinished: True, finished: False.
DEBUG 04-12 18:59:26 [core.py:414] EngineCore waiting for work.
INFO:     127.0.0.1:58276 - "POST /v1/completions HTTP/1.1" 400 Bad Request
INFO:     127.0.0.1:58282 - "POST /v1/completions/ HTTP/1.1" 307 Temporary Redirect
DEBUG 04-12 18:59:27 [core.py:420] EngineCore loop active - local unfinished: True, finished: False.
DEBUG 04-12 18:59:27 [core.py:414] EngineCore waiting for work.
INFO:     127.0.0.1:58282 - "POST /v1/completions HTTP/1.1" 200 OK
INFO:     127.0.0.1:58294 - "POST /v1/completions/ HTTP/1.1" 307 Temporary Redirect
WARNING 04-12 18:59:27 [protocol.py:70] The following fields were present in the request but ignored: {'prompts'}
DEBUG 04-12 18:59:27 [core.py:420] EngineCore loop active - local unfinished: True, finished: False.
DEBUG 04-12 18:59:27 [core.py:414] EngineCore waiting for work.
INFO:     127.0.0.1:58294 - "POST /v1/completions HTTP/1.1" 400 Bad Request
```

[prompts.json](https://github.com/user-attachments/files/19721599/prompts.json)

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by DarkLight1337** [comment]
cc @patrickvonplaten @njhill

**Comment by sam-paech** [comment]
I'm running into the same issue with the recently released Mistral small 3.2, when requesting top logprobs.

I'm really hoping this can be fixed as my use case depends on logprobs!

```
2025-06-21 06:33:41,337 [ERROR] [api_client.api_clien]: Error detail: {'object': 'error', 'message': "4 validation errors for CompletionLogProbs\ntop_logprobs.9.b'\\xe2\\x80'.[key]\n  Input should be a valid string, unable to parse raw data as a unicode string [type=string_unicode, input_value=b'\\xe2\\x80', input_type=bytes]\n    For further information visit https://errors.pydantic.dev/2.11/v/string_unicode\ntop_logprobs.9.b'\\xc2'.[key]\n  Input should be a valid string, unable to parse raw data as a unicode string [type=string_unicode, input_value=b'\\xc2', input_type=bytes]\n    For further information visit https://errors.pydantic.dev/2.11/v/string_unicode\ntop_logprobs.166.b'\\xc2'.[key]\n  Input should be a valid string, unable to parse raw data as a unicode string [type=string_unicode, input_value=b'\\xc2', input_type=bytes]\n    For further information visit https://errors.pydantic.dev/2.11/v/string_unicode\ntop_logprobs.311.b'\\xe5\\x94'.[key]\n  Input should be a valid string, unable to parse raw data as a unicode string [type=string_unicode, input_value=b'\\xe5\\x94', input_type=bytes]\n    For further information visit https://errors.pydantic.dev/2.11/v/string_unicode", 'type': 'BadRequestError', 'param': None, 'code': 400}
2025-06-21 06:33:41,337 [ERROR] [core.sampler        ]: API call failed: 400 Client Error: Bad Request for url: http://127.0.0.1:8000/v1/completions
```

**Comment by LimLims** [comment]
I found a workaround for this issue, at least for mistral small 3.2:

set this env var before running vllm:

`export VLLM_USE_V1=0`

**Comment by sam-paech** [comment]
For mistral small 3.2, I found using VLLM engine v0 is an effective workaround for this issue:

`export VLLM_USE_V1=0`
